pkg/controller/machine/drain_controller.go

```diff
@@ -0,0 +1,163 @@
+	client.Client
+	config *rest.Config
+	scheme *runtime.Scheme
+
+	eventRecorder record.EventRecorder
+}
+
+// newDrainController returns a new reconcile.Reconciler for machine-drain-controller
+func newDrainController(mgr manager.Manager) reconcile.Reconciler {
+	d := &machineDrainController{
+		Client:        mgr.GetClient(),
+		eventRecorder: mgr.GetEventRecorderFor("machine-drain-controller"),
+		config:        mgr.GetConfig(),
+		scheme:        mgr.GetScheme(),
+	}
+	return d
+}
+
+func (d *machineDrainController) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	m := &machinev1.Machine{}
+	if err := d.Client.Get(ctx, request.NamespacedName, m); err != nil {
+		if apierrors.IsNotFound(err) {
+			// Object not found, return.
+			return reconcile.Result{}, nil
+		}
+
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, err
+	}
+
+	existingDrainedCondition := conditions.Get(m, machinev1.MachineDrained)
+	alreadyDrained := existingDrainedCondition != nil && existingDrainedCondition.Status == corev1.ConditionTrue
+
+	if !m.ObjectMeta.DeletionTimestamp.IsZero() && *m.Status.Phase == phaseDeleting && !alreadyDrained {
+		drainFinishedCondition := conditions.TrueCondition(machinev1.MachineDrained)
+
+		if _, exists := m.ObjectMeta.Annotations[ExcludeNodeDrainingAnnotation]; !exists && m.Status.NodeRef != nil {
+			// pre-drain.delete lifecycle hook
+			// Return early without error, will requeue if/when the hook owner removes the annotation.
+			if len(m.Spec.LifecycleHooks.PreDrain) > 0 {
+				klog.Infof("%v: not draining machine: lifecycle blocked by pre-drain hook", m.Name)
+				d.eventRecorder.Eventf(m, corev1.EventTypeNormal, "DrainBlocked", "Drain blocked by pre-drain hook")
+				return reconcile.Result{}, nil
+			}
+			d.eventRecorder.Eventf(m, corev1.EventTypeNormal, "DrainProceeds", "Node drain proceeds")
+			if err := d.drainNode(ctx, m); err != nil {
+				klog.Errorf("%v: failed to drain node for machine: %v", m.Name, err)
+				conditions.Set(m, conditions.FalseCondition(
+					machinev1.MachineDrained,
+					machinev1.MachineDrainError,
+					machinev1.ConditionSeverityWarning,
+					"could not drain machine: %v", err,
+				))
+				d.eventRecorder.Eventf(m, corev1.EventTypeNormal, "DrainRequeued", "Node drain requeued: %v", err.Error())
+				return delayIfRequeueAfterError(err)
+			}
+			d.eventRecorder.Eventf(m, corev1.EventTypeNormal, "DrainSucceeded", "Node drain succeeded")
+			drainFinishedCondition.Message = "Drain finished successfully"
+		} else {
+			d.eventRecorder.Eventf(m, corev1.EventTypeNormal, "DrainSkipped", "Node drain skipped")
+			drainFinishedCondition.Message = "Node drain skipped"
+		}
+
+		d.eventRecorder.Eventf(m, corev1.EventTypeNormal, "DrainFinished", "Node drain finished")
+		conditions.Set(m, drainFinishedCondition)
+		// requeue request in case of failed update
+		err := d.Client.Status().Update(ctx, m)
+		return reconcile.Result{}, err
```

Comment:
Not a fan of naked error returns like this, can we wrap it with context, adding unique context makes it easier to find where it came from when debugging
```suggestion
		if err := d.Client.Status().Update(ctx, m); err != nil {
		  return reconcile.Result{}, fmt.Errorf("could not update machine status: %w", err)
		}
		
		return reconcile.Result{}, nil
```